sendrecv/gst-rust/src/main.rs

```diff
@@ -0,0 +1,768 @@
+            let mut app_control = self.0.lock().unwrap();
+            if app_control.app_state != AppState::PeerConnecting {
+                return Err(OutOfOrder("Received SESSION_OK when not calling"))?;
+            }
+            app_control.app_state = AppState::PeerConnected;
+        }
+        self.start_pipeline()
+    }
+    fn handle_error(&self) -> Result<(), Error> {
+        let app_control = self.0.lock().unwrap();
+        let error = match app_control.app_state {
+            AppState::ServerRegistering => AppState::ServerRegisteringError,
+            AppState::PeerConnecting => AppState::PeerConnectionError,
+            AppState::PeerConnected => AppState::PeerCallError,
+            AppState::PeerCallNegotiating => AppState::PeerCallError,
+            AppState::ServerRegisteringError => AppState::ServerRegisteringError,
+            AppState::PeerConnectionError => AppState::PeerConnectionError,
+            AppState::PeerCallError => AppState::PeerCallError,
+            AppState::AppStateErr => AppState::AppStateErr,
+            AppState::ServerConnected => AppState::AppStateErr,
+            AppState::ServerRegistered => AppState::AppStateErr,
+            AppState::PeerCallStarted => AppState::AppStateErr,
+        };
+        return Err(WsError(error))?;
+    }
+    fn handle_sdp(&self, type_: String, sdp: String) {
+        if !self.app_state_eq(AppState::PeerCallNegotiating, "Not ready to handle sdp") {
+            return;
+        }
+
+        if type_ != "answer" {
+            self.send_bus_error(String::from("Sdp type is not \"anser\""));
+            return;
+        }
+
+        let mut app_control = self.0.lock().unwrap();
+        print!("Received answer:\n{}\n", sdp);
+
+        let ret = gst_sdp::SDPMessage::parse_buffer(sdp.as_bytes()).unwrap();
+        let answer =
+            gst_webrtc::WebRTCSessionDescription::new(gst_webrtc::WebRTCSDPType::Answer, ret);
+        let promise = gst::Promise::new();
+        app_control
+            .webrtc
+            .as_ref()
+            .unwrap()
+            .emit("set-remote-description", &[&answer, &promise])
+            .unwrap();
+        app_control.app_state = AppState::PeerCallStarted;
+    }
+    fn handle_ice(&self, sdp_mline_index: u32, candidate: String) {
+        let app_control = self.0.lock().unwrap();
+        app_control
+            .webrtc
+            .as_ref()
+            .unwrap()
+            .emit("add-ice-candidate", &[&sdp_mline_index, &candidate])
+            .unwrap();
+    }
+    fn on_message(&mut self, msg: String) -> Result<(), Error> {
+        if msg == "HELLO" {
+            return self.handle_hello();
+        }
+        if msg == "SESSION_OK" {
+            return self.handle_session_ok();
+        }
+
+        if msg.starts_with("ERROR") {
+            println!("Got error message! {}", msg);
+            return self.handle_error();
+        }
+        let json_msg: JsonMsg = serde_json::from_str(&msg).unwrap();
```

Comment:
Instead of unwrapping, failing to deserialize the JSON should probably just be an error like other errors